packages/data-provider/src/utils/submittedDataUtils.ts

```diff
@@ -16,15 +17,74 @@ import {
 import { BaseDependencies } from '../config/config.js';
 import {
 	DataRecordReference,
-	MERGE_REFERENCE_TYPE,
 	type GroupedDataSubmission,
+	MERGE_REFERENCE_TYPE,
+	type MutableDataDiff,
 	type SubmittedDataResponse,
 } from './types.js';
 
 const utils = (dependencies: BaseDependencies) => {
 	const LOG_MODULE = 'SUBMITTED_DATA_UTILS';
 	const { logger } = dependencies;
 	return {
+		/**
+		 * Compares two `DataRecord` objects and returns the differences between them.
+		 * @param oldRecord The original `DataRecord` object to compare.
+		 * @param newRecord The new `DataRecord` object to compare against the original.
+		 * @returns An object of type `DataDiff` containing the differences between `oldRecord` and `newRecord`.
+		 * 	The differing values are recorded with the `old` object containing the values
+		 * 	from `oldRecord` and the `new` object containing the corresponding values from `newRecord`.
+		 */
+		computeDataDiff: (oldRecord: DataRecord | null, newRecord: DataRecord | null): DataDiff => {
```

Comment:
We should write unit tests for this:
1. returns no change between 2 null values
2. returns no change between 2 objects with same values
3. reports new fields added to newRecord
4. reports fields removed from newRecord
5. reports values changed from oldRecord to newRecord
6. returns whole newRecord when compared to null oldRecord
7. returns whole oldRecord deleted when compared to null newRecord

Reply:
yes definitely need to add these tests. Thanks

Reply:
Test cases were added to cover these and other scenearios